compiler/plugins/target/ROCM/test/opt_pass_plugin/CMakeLists.txt

```diff
@@ -0,0 +1,19 @@
+iree_cc_library(
+  NAME
+    GPUHello
+  SRCS
+    "GPUHello.cpp"
+  DEPS
+    LLVMSupport
```

Comment:
doesn't this also depend on the library that exposes 'core llvm' (IR apis)

Reply:
Well, it depends. If you're compiling it to a shared object that gets loaded at run/JIT compile time you have to be careful about getting duplicate symbols from the LLVM libs. So the LLVMCore module usually gets linked to the compiler and just comes along when the plugin is loaded. The other option is to try and have CMake link these as `PRIVATE` but I have found that be to hit or miss.

Reply:
I also observed this issue with LLVMcore that will segfault during dlopen, because of the duplicated symbols in the CLI.

Reply:
> The `PkgCI / Build Packages / Linux Release (x86_64)` error is odd to me:
> 
> ```
> compiler/plugins/target/ROCM/test/opt_pass_plugin/CMakeFiles/iree_compiler_plugins_target_ROCM_test_opt_pass_plugin_GPUHello.objects.dir/GPUHello.cpp.o
>   FAILED: compiler/plugins/target/ROCM/test/opt_pass_plugin/CMakeFiles/iree_compiler_plugins_target_ROCM_test_opt_pass_plugin_GPUHello.objects.dir/GPUHello.cpp.o
>   ccache /usr/bin/clang++  -I/home/runner/_work/iree/iree -I/home/runner/_work/iree/iree/compiler/build/b -I/home/runner/_work/iree/iree/third_party/llvm-project/llvm/include -I/home/runner/_work/iree/iree/compiler/build/b/llvm-project/include -I/home/runner/_work/iree/iree/third_party/llvm-project/mlir/include -I/home/runner/_work/iree/iree/compiler/build/b/llvm-project/tools/mlir/include -I/home/runner/_work/iree/iree/third_party/llvm-project/lld/include -I/home/runner/_work/iree/iree/compiler/build/b/llvm-project/tools/lld/include -O3 -DNDEBUG -std=gnu++17 -fPIC -fvisibility=hidden -fno-rtti -fno-exceptions -Werror -Wall -Wno-error=deprecated-declarations -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wno-missing-braces -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wpointer-arith -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -fno-lax-vector-conversions -fmacro-prefix-map=/home/runner/_work/iree/iree=iree -pthread -MD -MT compiler/plugins/target/ROCM/test/opt_pass_plugin/CMakeFiles/iree_compiler_plugins_target_ROCM_test_opt_pass_plugin_GPUHello.objects.dir/GPUHello.cpp.o -MF compiler/plugins/target/ROCM/test/opt_pass_plugin/CMakeFiles/iree_compiler_plugins_target_ROCM_test_opt_pass_plugin_GPUHello.objects.dir/GPUHello.cpp.o.d -o compiler/plugins/target/ROCM/test/opt_pass_plugin/CMakeFiles/iree_compiler_plugins_target_ROCM_test_opt_pass_plugin_GPUHello.objects.dir/GPUHello.cpp.o -c /home/runner/_work/iree/iree/compiler/plugins/target/ROCM/test/opt_pass_plugin/GPUHello.cpp
>   In file included from /home/runner/_work/iree/iree/compiler/plugins/target/ROCM/test/opt_pass_plugin/GPUHello.cpp:7:
>   In file included from /home/runner/_work/iree/iree/third_party/llvm-project/llvm/include/llvm/IR/IRBuilder.h:24:
>   In file included from /home/runner/_work/iree/iree/third_party/llvm-project/llvm/include/llvm/IR/ConstantFolder.h:21:
>   In file included from /home/runner/_work/iree/iree/third_party/llvm-project/llvm/include/llvm/IR/ConstantFold.h:24:
>   In file included from /home/runner/_work/iree/iree/third_party/llvm-project/llvm/include/llvm/IR/InstrTypes.h:24:
>   /home/runner/_work/iree/iree/third_party/llvm-project/llvm/include/llvm/IR/Attributes.h:90:14: fatal error: 'llvm/IR/Attributes.inc' file not found
>      90 |     #include "llvm/IR/Attributes.inc"
>         |              ^~~~~~~~~~~~~~~~~~~~~~~~
>   1 error generated.
> ```
> 
> This is within the llvm project, I don't see how what I am doing can break this.

I think the error logs here are similar to Jakub's question:

https://github.com/iree-org/iree/actions/runs/11060154642/job/30730692896?pr=18347#step:7:3846

This library needs other LLVM targets added to `DEPS` for it to build reliably.

If this target also built with Bazel, that would give similar errors specific to each header file that is missing a dep. Bazel support for tests would be a nice bonus, but I think you should be able to work through this error with just CMake. One thing you can try is start with a clean build directory then _just_ build the `GPUHello` target. That way, other CMake target builds won't incidentally build the other LLVM targets that this should be depending on first.

Reply:
If duplicated symbols is an issue, can this test shared object be built with hidden symbols?

Reply:
You may be able to get away with something like setting the upper level CMake file symbols to hidden and resetting it here
`target_compile_options(GPUHello PRIVATE -fvisibility=default)`
This is not an uncommon approach in frameworks that mix and match LLVM libs do this.